File: test-sdks-harness/java/AcIt.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.svix.AutoConfig;
import com.svix.AutoConfigConsumer;
import com.svix.exceptions.ApiException;
import com.svix.internalapi.MessagePollerv2ConsumerPollOptions;
import com.svix.models.DestinationOut;
import com.svix.models.EndpointIn;
import com.svix.models.EndpointOut;
import com.svix.models.PollerV2MessageOut;
import com.svix.models.PollerV2PollOut;
import com.svix.models.SinkInCommon;
import com.svix.models.StartingPosition;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AcIt {
    public static void main(String[] args) throws Exception {
        String path = System.getenv("AUTOCONFIG_FIXTURES");
        JsonNode root = new ObjectMapper().readTree(Files.readString(Path.of(path)));
        JsonNode lang = root.get("languages").get("java");
        String serverUrl = root.get("serverUrl").asText();
        String orgToken = root.get("orgToken").asText();
        String eventType = root.get("eventType").asText();
        String httpUrl = root.get("httpUrl").asText();
        String consumerId = root.get("consumerId").asText();
        String appId = lang.get("appId").asText();

        System.out.println("LANG: java");
        System.out.println(runHttp("A", lang.get("v1Http").asText(), httpUrl, eventType));
        System.out.println(runHttp("B", lang.get("v2Http").asText(), httpUrl, eventType));
        System.out.println(
                runPoller(
                        "C",
                        lang.get("v1Poller").asText(),
                        eventType,
                        consumerId,
                        serverUrl,
                        orgToken,
                        appId,
                        "java-v1"));
        System.out.println(
                runPoller(
                        "D",
                        lang.get("v2Poller").asText(),
                        eventType,
                        consumerId,
                        serverUrl,
                        orgToken,
                        appId,
                        "java-v2"));
        System.out.println(
                runPollerExisting(
                        "E",
                        lang.get("v2PollerExisting").asText(),
                        eventType,
                        consumerId,
                        serverUrl,
                        orgToken,
                        appId,
                        "java-e"));
        System.out.println(
                "Notes: SinkInCommon with eventTypes; no filled optionals");
    }

    private static String runHttp(String label, String token, String httpUrl, String eventType) {
        try {
            EndpointIn endpoint =
                    new EndpointIn()
                            .url(URI.create(httpUrl))
                            .eventTypes(new LinkedHashSet<>(Set.of(eventType)));
            EndpointOut out = new AutoConfig(token, endpoint).subscribe();
            if (out == null || out.getId() == null || out.getId().isEmpty()) {
                return label + ": FAIL — subscribe returned empty id";
            }
            if (out.getUrl() == null || !httpUrl.equals(out.getUrl().toString())) {
                return label + ": FAIL — url mismatch got=" + out.getUrl() + " id=" + out.getId();
            }
            return label + ": PASS — id=" + out.getId() + " url=" + out.getUrl();
        } catch (ApiException e) {
            return label + ": FAIL — ApiException " + e.getCode() + " " + e.getMessage() + " "
                    + e.getResponseBody();
        } catch (Exception e) {
            return label + ": FAIL — " + e.getClass().getSimpleName() + ": " + e.getMessage();
        }
    }

    private static SinkInCommon minimalSink(String eventType) {
        return new SinkInCommon().eventTypes(Set.of(eventType));
    }

    private static String runPoller(
            String label,
            String token,
            String eventType,
            String consumerId,
            String serverUrl,
            String orgToken,
            String appId,
            String src) {
        DestinationOut dest;
        AutoConfigConsumer consumer;
        try {
            consumer = new AutoConfigConsumer(token, minimalSink(eventType));
            dest = consumer.subscribe();
        } catch (ApiException e) {
            if (e.getCode() == 501) {
                return label + ": FAIL — DIOM_MISSING " + e.getResponseBody();
            }
            return label + ": FAIL — subscribe ApiException " + e.getCode() + " " + e.getMessage()
                    + " " + e.getResponseBody();
        } catch (Exception e) {
            return label + ": FAIL — subscribe " + e.getClass().getSimpleName() + ": "
                    + e.getMessage();
        }

        if (dest == null || dest.getId() == null || dest.getId().isEmpty()) {
            return label + ": FAIL — subscribe returned empty dest id";
        }

        try {
            sendMsg(serverUrl, orgToken, appId, eventType, src);
        } catch (Exception e) {
            return label + ": FAIL — send msg " + e.getClass().getSimpleName() + ": "
                    + e.getMessage();
        }

        MessagePollerv2ConsumerPollOptions opts = new MessagePollerv2ConsumerPollOptions();
        opts.setStartingPosition(StartingPosition.EARLIEST);
        opts.setLeaseDurationMs(2000L);

        PollerV2MessageOut hit = null;
        String lastErr = null;
        for (int i = 0; i < 10; i++) {
            try {
                PollerV2PollOut poll = consumer.receive(consumerId, opts);
                if (poll != null && poll.getData() != null) {
                    for (PollerV2MessageOut msg : poll.getData()) {
                        if (eventType.equals(msg.getEventType()) && srcEquals(msg.getPayload(), src)) {
                            hit = msg;
                            break;
                        }
                    }
                }
                if (hit != null) {
                    break;
                }
                lastErr = "empty/no-match";
            } catch (ApiException e) {
                if (e.getCode() == 501) {
                    return label + ": FAIL — DIOM_MISSING dest=" + dest.getId() + " "
                            + e.getResponseBody();
                }
                lastErr = "receive ApiException " + e.getCode() + " " + e.getResponseBody();
            } catch (Exception e) {
                lastErr = "receive " + e.getClass().getSimpleName() + ": " + e.getMessage();
            }
            try {
                Thread.sleep(2000);
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
                return label + ": FAIL — interrupted";
            }
        }

        if (hit == null) {
            return label + ": FAIL — dest=" + dest.getId()
                    + " no matching message after 10x2s; last=" + lastErr;
        }

        long offset = hit.getOffset();
        try {
            consumer.commit(consumerId, offset);
        } catch (Exception e) {
            return label + ": FAIL — dest=" + dest.getId() + " commit offset=" + offset + " "
                    + e.getClass().getSimpleName() + ": " + e.getMessage();
        }

        try {
            PollerV2PollOut after = consumer.receive(consumerId, opts);
            if (after != null && after.getData() != null) {
                for (PollerV2MessageOut msg : after.getData()) {
                    if (msg.getOffset() != null && msg.getOffset() == offset) {
                        return label + ": FAIL — dest=" + dest.getId() + " committed offset "
                                + offset + " came back";
                    }
                }
            }
        } catch (Exception e) {
            return label + ": FAIL — dest=" + dest.getId() + " post-commit receive "
                    + e.getClass().getSimpleName() + ": " + e.getMessage();
        }

        return label + ": PASS — dest=" + dest.getId() + " offset=" + offset + " src=" + src;
    }

    private static String runPollerExisting(
            String label,
            String token,
            String eventType,
            String consumerId,
            String serverUrl,
            String orgToken,
            String appId,
            String src) {
        DestinationOut dest;
        AutoConfigConsumer consumer;
        try {
            AutoConfigConsumer binder = new AutoConfigConsumer(token, minimalSink(eventType));
            dest = binder.subscribe();
            if (dest == null || dest.getId() == null || dest.getId().isEmpty()) {
                return label + ": FAIL — subscribe returned empty dest id";
            }
            consumer = new AutoConfigConsumer(token, minimalSink(eventType));
        } catch (ApiException e) {
            if (e.getCode() == 501) {
                return label + ": FAIL — DIOM_MISSING " + e.getResponseBody();
            }
            return label + ": FAIL — subscribe ApiException " + e.getCode() + " " + e.getMessage()
                    + " " + e.getResponseBody();
        } catch (Exception e) {
            return label + ": FAIL — " + e.getClass().getSimpleName() + ": " + e.getMessage();
        }

        try {
            sendMsg(serverUrl, orgToken, appId, eventType, src);
        } catch (Exception e) {
            return label + ": FAIL — send msg " + e.getClass().getSimpleName() + ": "
                    + e.getMessage();
        }

        MessagePollerv2ConsumerPollOptions opts = new MessagePollerv2ConsumerPollOptions();
        opts.setStartingPosition(StartingPosition.EARLIEST);
        opts.setLeaseDurationMs(2000L);

        PollerV2MessageOut hit = null;
        String lastErr = null;
        for (int i = 0; i < 10; i++) {
            try {
                PollerV2PollOut poll = consumer.receive(consumerId, opts);
                if (poll != null && poll.getData() != null) {
                    for (PollerV2MessageOut msg : poll.getData()) {
                        if (eventType.equals(msg.getEventType()) && srcEquals(msg.getPayload(), src)) {
                            hit = msg;
                            break;
                        }
                    }
                }
                if (hit != null) {
                    break;
                }
                lastErr = "empty/no-match";
            } catch (ApiException e) {
                if (e.getCode() == 501) {
                    return label + ": FAIL — DIOM_MISSING " + e.getResponseBody();
                }
                lastErr = "receive ApiException " + e.getCode() + " " + e.getResponseBody();
            } catch (Exception e) {
                lastErr = "receive " + e.getClass().getSimpleName() + ": " + e.getMessage();
            }
            try {
                Thread.sleep(2000);
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
                return label + ": FAIL — interrupted";
            }
        }

        if (hit == null) {
            return label + ": FAIL — no matching message after 10x2s; last=" + lastErr;
        }

        long offset = hit.getOffset();
        try {
            consumer.commit(consumerId, offset);
        } catch (Exception e) {
            return label + ": FAIL — commit offset=" + offset + " "
                    + e.getClass().getSimpleName() + ": " + e.getMessage();
        }

        try {
            PollerV2PollOut after = consumer.receive(consumerId, opts);
            if (after != null && after.getData() != null) {
                for (PollerV2MessageOut msg : after.getData()) {
                    if (msg.getOffset() != null && msg.getOffset() == offset) {
                        return label + ": FAIL — committed offset " + offset + " came back";
                    }
                }
            }
        } catch (Exception e) {
            return label + ": FAIL — post-commit receive "
                    + e.getClass().getSimpleName() + ": " + e.getMessage();
        }

        return label + ": PASS — dest=" + dest.getId() + " offset=" + offset + " src=" + src
                + " (no subscribe on receiver)";
    }

    private static void sendMsg(
            String serverUrl, String orgToken, String appId, String eventType, String src)
            throws Exception {
        String body =
                "{\"eventType\":\""
                        + eventType
                        + "\",\"payload\":{\"ok\":true,\"src\":\""
                        + src
                        + "\"}}";
        HttpRequest req =
                HttpRequest.newBuilder()
                        .uri(URI.create(serverUrl + "/api/v1/app/" + appId + "/msg/"))
                        .header("Authorization", "Bearer " + orgToken)
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                        .build();
        HttpResponse<String> resp =
                HttpClient.newHttpClient().send(req, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() < 200 || resp.statusCode() >= 300) {
            throw new RuntimeException("HTTP " + resp.statusCode() + " " + resp.body());
        }
    }

    private static boolean srcEquals(Object payload, String src) {
        if (payload instanceof Map) {
            Object v = ((Map<?, ?>) payload).get("src");
            return src.equals(String.valueOf(v));
        }
        return payload != null && payload.toString().contains("\"src\":\"" + src + "\"");
    }
}

File: test-sdks-harness/kotlin/AcIt.kt
package com.svix.kotlin

import com.svix.kotlin.exceptions.ApiException
import com.svix.kotlin.internal.MessagePollerv2ConsumerPollOptions
import com.svix.kotlin.models.EndpointIn
import com.svix.kotlin.models.SinkInCommon
import com.svix.kotlin.models.StartingPosition
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlin.test.Test
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

class AcItHarness {
    @Serializable
    private data class LangFixtures(
        val appId: String,
        val v1Http: String,
        val v1Poller: String,
        val v2Http: String,
        val v2Poller: String,
        val v2PollerExisting: String,
    )

    @Serializable
    private data class FixturesFile(
        val serverUrl: String,
        val orgToken: String,
        val eventType: String,
        val httpUrl: String,
        val consumerId: String,
        val languages: Map<String, LangFixtures>,
    )

    @Test
    fun liveAutoConfig() {
        val json = Json { ignoreUnknownKeys = true }
        val fixturesPath = System.getenv("AUTOCONFIG_FIXTURES")
            ?: error("AUTOCONFIG_FIXTURES is unset")
        val fixtures =
            json.decodeFromString<FixturesFile>(java.io.File(fixturesPath).readText())
        val lang = fixtures.languages["kotlin"]!!
        val results = LinkedHashMap<String, String>()
        val endpointIn =
            EndpointIn(url = fixtures.httpUrl, eventTypes = setOf(fixtures.eventType))

        results["A"] =
            runCase("v1 HTTP") {
                val out = runBlocking { AutoConfig(lang.v1Http, endpointIn).subscribe() }
                check(out.url == fixtures.httpUrl) {
                    "url mismatch: ${out.url} != ${fixtures.httpUrl}"
                }
                "id=${out.id} url=${out.url}"
            }

        results["B"] =
            runCase("v2 HTTP") {
                val out = runBlocking { AutoConfig(lang.v2Http, endpointIn).subscribe() }
                check(out.url == fixtures.httpUrl) {
                    "url mismatch: ${out.url} != ${fixtures.httpUrl}"
                }
                "id=${out.id} url=${out.url}"
            }

        results["C"] =
            runCase("v1 poller") {
                runPoller(
                    token = lang.v1Poller,
                    src = "kotlin-v1",
                    fixtures = fixtures,
                    appId = lang.appId,
                )
            }

        results["D"] =
            runCase("v2 poller") {
                runPoller(
                    token = lang.v2Poller,
                    src = "kotlin-v2",
                    fixtures = fixtures,
                    appId = lang.appId,
                )
            }

        results["E"] =
            runCase("v2 receive without subscribe") {
                runPollerExisting(
                    token = lang.v2PollerExisting,
                    src = "kotlin-e",
                    fixtures = fixtures,
                    appId = lang.appId,
                )
            }

        val report =
            buildString {
                appendLine("LANG: kotlin")
                for (key in listOf("A", "B", "C", "D", "E")) {
                    appendLine("$key: ${results[key]}")
                }
                appendLine(
                    "Notes: copied into kotlin test source set for this run; JDK 17"
                )
            }
        val outPath = System.getenv("AUTOCONFIG_RESULTS")
        if (outPath != null) {
            java.io.File(outPath).writeText(report)
        }
        print(report)
    }

    private fun runCase(label: String, block: () -> String): String {
        return try {
            val detail = block()
            "PASS — $label $detail"
        } catch (e: ApiException) {
            val body = e.body?.replace('\n', ' ')?.take(240) ?: ""
            "FAIL — $label ${e.message} $body".trim()
        } catch (e: Throwable) {
            val msg = e.message?.replace('\n', ' ') ?: e.javaClass.simpleName
            "FAIL — $label $msg"
        }
    }

    private fun runPoller(
        token: String,
        src: String,
        fixtures: FixturesFile,
        appId: String,
    ): String {
        return runBlocking {
            val consumer =
                AutoConfigConsumer(
                    token,
                    SinkInCommon(eventTypes = setOf(fixtures.eventType)),
                )
            val dest = consumer.subscribe()
            sendMsg(fixtures, appId, src)

            val options =
                MessagePollerv2ConsumerPollOptions(
                    startingPosition = StartingPosition.EARLIEST,
                    leaseDurationMs = 2000UL,
                )

            var foundOffset: ULong? = null
            var lastCount = 0
            for (attempt in 1..10) {
                val poll = consumer.receive(fixtures.consumerId, options)
                lastCount = poll.data.size
                val match =
                    poll.data.firstOrNull { msg ->
                        msg.eventType == fixtures.eventType &&
                            (msg.payload["src"]?.toString() == src)
                    }
                if (match != null) {
                    foundOffset = match.offset
                    break
                }
                delay(2000)
            }
            check(foundOffset != null) {
                "no message with src=$src after 10 retries (last poll size=$lastCount)"
            }

            consumer.commit(fixtures.consumerId, foundOffset!!)

            val after = consumer.receive(fixtures.consumerId, options)
            val replayed = after.data.any { it.offset == foundOffset }
            check(!replayed) { "committed offset $foundOffset was replayed" }

            "dest=${dest.id} offset=$foundOffset"
        }
    }

    private fun runPollerExisting(
        token: String,
        src: String,
        fixtures: FixturesFile,
        appId: String,
    ): String {
        return runBlocking {
            val binder =
                AutoConfigConsumer(
                    token,
                    SinkInCommon(eventTypes = setOf(fixtures.eventType)),
                )
            val dest = binder.subscribe()
            val consumer =
                AutoConfigConsumer(
                    token,
                    SinkInCommon(eventTypes = setOf(fixtures.eventType)),
                )
            sendMsg(fixtures, appId, src)

            val options =
                MessagePollerv2ConsumerPollOptions(
                    startingPosition = StartingPosition.EARLIEST,
                    leaseDurationMs = 2000UL,
                )

            var foundOffset: ULong? = null
            var lastCount = 0
            for (attempt in 1..10) {
                val poll = consumer.receive(fixtures.consumerId, options)
                lastCount = poll.data.size
                val match =
                    poll.data.firstOrNull { msg ->
                        msg.eventType == fixtures.eventType &&
                            (msg.payload["src"]?.toString() == src)
                    }
                if (match != null) {
                    foundOffset = match.offset
                    break
                }
                delay(2000)
            }
            check(foundOffset != null) {
                "no message with src=$src after 10 retries (last poll size=$lastCount)"
            }

            consumer.commit(fixtures.consumerId, foundOffset!!)

            val after = consumer.receive(fixtures.consumerId, options)
            val replayed = after.data.any { it.offset == foundOffset }
            check(!replayed) { "committed offset $foundOffset was replayed" }

            "dest=${dest.id} offset=$foundOffset no subscribe on receiver"
        }
    }

    private fun sendMsg(fixtures: FixturesFile, appId: String, src: String) {
        val body =
            """{"eventType":"${fixtures.eventType}","payload":{"ok":true,"src":"$src"}}"""
        val req =
            HttpRequest.newBuilder()
                .uri(URI.create("${fixtures.serverUrl}/api/v1/app/$appId/msg/"))
                .header("Authorization", "Bearer ${fixtures.orgToken}")
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build()
        val resp = HttpClient.newHttpClient().send(req, HttpResponse.BodyHandlers.ofString())
        check(resp.statusCode() in 200..299) {
            "send msg status=${resp.statusCode()} body=${resp.body()}"
        }
    }
}
